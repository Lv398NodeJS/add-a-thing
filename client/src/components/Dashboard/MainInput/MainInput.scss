$High: red;
$Medium: orange;
$Low: green;

.inputicon{
  width: 1.7rem;
}

.btn-outline-primary:hover {
  background-color: #74aeb0;
  border-color: #74aeb0;
}

.btn-outline-primary {
  border-color: #74aeb0;
}

.main-input .form-control:focus{
  box-shadow: none;
}

.main-input.input-group input.form-control{
  border-right: none;
}

.main-input.input-group input.form-control + select.form-control{
  margin-left: 0;
  width: 56px;
  text-align-last: center;
  flex-grow: unset;
  -moz-appearance: none;
  -webkit-appearance: none;
}
.main-input.input-group option{
  text-align-last: center;
}

.error.form-control{
  border-color: $High;
}

.main-input.input-group input.form-control + select.form-control.priorityH {
  border-left: 2px solid $High;
}

.main-input.input-group input.form-control + select.form-control.priorityM {
  border-left: 2px solid $Medium;
}

.main-input.input-group input.form-control + select.form-control.priorityL {
  border-left: 2px solid $Low;
}

.gray-tooltip.tooltip .tooltip-inner{
  background: rgba(212, 212, 212, 1);
  padding: 2px 10px ;
  color: #fff;
  border-radius: 3px ;
  max-width: 100%;
}

@media (max-width: 576px){
  .main-input .input-group-append button:nth-last-child(1),.main-input .input-group-append button:nth-last-child(2){
    display: none;
  }
  .gray-tooltip.tooltip{
    margin-left: 10px;
  }
  .gray-tooltip.tooltip .tooltip-inner{
    max-width: calc(100% - 22px);
    margin: 0 auto;
  }
}
