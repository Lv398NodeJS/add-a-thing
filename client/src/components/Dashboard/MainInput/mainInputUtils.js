export const addTaskWithHash = inputData => inputData.slice(0, inputData.length - 2).trim();

export const showPriorityColor = (newPriority) => {
  let style = '';
  switch (newPriority) {
    case 'High':
      style += 'priorityH';
      break;
    case 'Medium':
      style += 'priorityM';
      break;
    case 'Low':
      style += 'priorityL';
      break;
    default:
      break;
  }
  return style;
};

export const getPriority = (enteredPriority) => {
  let newPriority = '';
  if (enteredPriority === '1') {
    newPriority = 'High';
  } else if (enteredPriority === '3') {
    newPriority = 'Low';
  } else {
    newPriority = 'Medium';
  }
  return newPriority;
};
