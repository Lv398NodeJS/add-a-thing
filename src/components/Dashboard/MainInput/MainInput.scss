$High: red;
$Medium: orange;
$Low: green;

.inputicon{
  width: 1.7rem;
}

.btn-outline-primary:hover {
  background-color: #74aeb0;
  border-color: #74aeb0;
}

.btn-outline-primary {
  border-color: #74aeb0;
}

.form-control:focus{
  box-shadow: none;
}

.input-group input.form-control{
  border-right: none;
}

.input-group input.form-control + select.form-control{
  margin-left: 0;
  width: 56px;
  text-align-last: center;
  flex-grow: unset;
  -moz-appearance: none;
  -webkit-appearance: none;
}
.input-group option{
  text-align-last: center;
}

.error.form-control{
  border-color: $High;
}

.input-group input.form-control + select.form-control.priorityH {
  border-left: 2px solid $High;
}

.input-group input.form-control + select.form-control.priorityM {
  border-left: 2px solid $Medium;
}

.input-group input.form-control + select.form-control.priorityL {
  border-left: 2px solid $Low;
}

.gray-tooltip.tooltip .tooltip-inner{
  background: rgba(212, 212, 212, 1);
  padding: 2px 10px ;
  color: #fff;
  border-radius: 3px ;
  max-width: 100%;
}

@media (max-width: 576px){
  .input-group-append button:nth-last-child(1),.input-group-append button:nth-last-child(2){
    display: none;
  }
  .gray-tooltip.tooltip{
    margin-left: 10px;
  }
  .gray-tooltip.tooltip .tooltip-inner{
    max-width: calc(100% - 22px);
    margin: 0 auto;
  }
}
