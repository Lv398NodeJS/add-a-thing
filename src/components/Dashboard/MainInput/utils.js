export const addTaskWithHash = inputData => inputData.slice(0, inputData.length - 2).trim();

export const showPriorityColor = (newPriority) => {
  let style = '';
  switch (newPriority) {
    case 'High':
      style += 'priorityH';
      break;
    case 'Medium':
      style += 'priorityM';
      break;
    case 'Low':
      style += 'priorityL';
      break;
    default:
      break;
  }
  return style;
};

export const changePriority = (getPriority) => {
  let setPriority = '';
  if (getPriority === '1') {
    setPriority = 'High';
  } else if (getPriority === '3') {
    setPriority = 'Low';
  } else {
    setPriority = 'Medium';
  }
  return setPriority;
};
