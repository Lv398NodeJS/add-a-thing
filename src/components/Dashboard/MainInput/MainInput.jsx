import React from 'react';
import {
  InputGroup,
  Button,
  FormControl,
  Form,
  OverlayTrigger,
  Tooltip,
} from 'react-bootstrap';
import add from '../../assets/add.svg';
import crossicon from '../../assets/crossicon.svg';
import './MainInput.scss';
import SpeechRecognition from '../SpeechRecognition/SpeechRecognition';

export default class MainInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      newTaskVal: '',
      newPriority: 'Medium',
      hasHash: false,
      error: false,
    };
  }

  showPriorityColor = () => {
    const { newPriority } = this.state;
    let style = '';
    switch (newPriority) {
      case 'High':
        style += 'priorityH';
        break;
      case 'Medium':
        style += 'priorityM';
        break;
      case 'Low':
        style += 'priorityL';
        break;
      default:
        break;
    }
    return style;
  }

  enterButtonPress = (button) => {
    const { newTaskVal, newPriority } = this.state;
    if (button.key !== 'Enter') return;
    this.sendNewTaskToParent(newTaskVal, newPriority);
  };

  clearInput = () => {
    this.setState({
      newTaskVal: '',
    });
  };

  setInputValue = (value) => {
    const { newPriority } = this.state;
    const inputDataArr = value.split('');
    const hashCheck = inputDataArr[inputDataArr.length - 2];
    let setPriority = newPriority;
    if (hashCheck === '#') {
      const getPriority = inputDataArr[inputDataArr.length - 1];
      if (getPriority === '1') {
        setPriority = 'High';
      } else if (getPriority === '3') {
        setPriority = 'Low';
      } else {
        setPriority = 'Medium';
      }
      this.setState({
        hasHash: true,
      });
    }
    this.setState({
      newTaskVal: inputDataArr.join(''),
      newPriority: setPriority,
    });
  };

  sendNewTaskToParent = (inputData, newPriority) => {
    const { addNewTask } = this.props;
    const { hasHash } = this.state;
    if (!inputData.trim().length) {
      this.setState({
        error: true,
      });
      return;
    }
    if (hasHash) {
      addNewTask(inputData.split('').slice(0, inputData.length - 2).join('').trim(), newPriority);
    } else {
      addNewTask(inputData.trim(), newPriority);
    }
    this.clearInput();
    this.setState({
      hasHash: false,
      error: false,
    });
  };

  handleSelect = (value) => {
    this.setState({
      newPriority: value,
    });
  }

  render() {
    const { newTaskVal, newPriority, error } = this.state;
    const returnError = error ? 'error' : '';

    return (
      <InputGroup className="mb-3 mt-3 main-input">
        <OverlayTrigger
          placement="bottom-start"
          trigger="focus"
          overlay={(
            <Tooltip className="gray-tooltip">Add #num in the end to set priority 1 High, 2 Medium, 3 Low</Tooltip>
          )}
        >
          <FormControl
            placeholder="Type task name"
            aria-label="Type task name"
            aria-describedby="basic-addon2"
            size="lg"
            maxLength="100"
            onChange={event => this.setInputValue(event.target.value)}
            onKeyPress={this.enterButtonPress}
            className={returnError}
            value={newTaskVal}
          />
        </OverlayTrigger>
        <OverlayTrigger
          placement="bottom"
          overlay={(
            <Tooltip>Click to set priority.</Tooltip>
          )}
        >
          <Form.Control
            as="select"
            size="lg"
            className={this.showPriorityColor()}
            value={newPriority}
            onChange={event => this.handleSelect(event.target.value)}
          >
            <option value="High">H</option>
            <option value="Medium">M</option>
            <option value="Low">L</option>
          </Form.Control>
        </OverlayTrigger>
        <InputGroup.Append>
          <SpeechRecognition setText={this.setInputValue} />
          <Button variant="outline-primary" onClick={() => this.sendNewTaskToParent(newTaskVal, newPriority)}>
            <img src={add} alt={add} className="inputicon" />
          </Button>
          <Button variant="outline-danger" onClick={this.clearInput}>
            <img src={crossicon} alt={crossicon} className="inputicon" />
          </Button>
        </InputGroup.Append>
      </InputGroup>
    );
  }
}
