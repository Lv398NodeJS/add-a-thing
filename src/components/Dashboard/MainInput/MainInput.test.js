import React from 'react';
import { shallow, mount } from 'enzyme';
import MainInput from './MainInput';

describe('MainInput tests', () => {
  it('MainInput should render correctly', () => {
    const component = shallow(<MainInput />);
    expect(component).toMatchSnapshot();
  });

  it('Should render 2 or more buttons', () => {
    const component = mount(<MainInput />);
    expect(component.find('button').length).not.toBeLessThan(2);
  });

  it('Should render 1 input', () => {
    const component = mount(<MainInput />);
    expect(component.find('input').length).toBe(1);
  });

  it('Should change state after input text', () => {
    const fakeVal = 'some text';
    const component = mount(<MainInput />);
    const input = component.find('input');
    input.simulate('change', {target: {value: fakeVal} });
    expect(component.state().newTaskVal).toEqual(fakeVal);
  });

});
