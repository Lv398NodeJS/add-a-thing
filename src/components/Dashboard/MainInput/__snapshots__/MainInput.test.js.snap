// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainInput tests MainInput should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainInput />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef
          aria-describedby="basic-addon2"
          aria-label="Type task name"
          as="input"
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="Type task name"
          size="lg"
          value=""
        />,
        <ForwardRef>
          <SpeechRecognition
            setText={[Function]}
          />
          <ForwardRef
            active={false}
            disabled={false}
            onClick={[Function]}
            type="button"
            variant="outline-primary"
          >
            <img
              alt="add.svg"
              className="inputicon"
              src="add.svg"
            />
          </ForwardRef>
          <ForwardRef
            active={false}
            disabled={false}
            onClick={[Function]}
            type="button"
            variant="outline-danger"
          >
            <img
              alt="crossicon.svg"
              className="inputicon"
              src="crossicon.svg"
            />
          </ForwardRef>
        </ForwardRef>,
      ],
      "className": "mb-3 mt-3",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "aria-describedby": "basic-addon2",
          "aria-label": "Type task name",
          "as": "input",
          "onChange": [Function],
          "onKeyPress": [Function],
          "placeholder": "Type task name",
          "size": "lg",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "Feedback": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "as": "div",
              "type": "valid",
            },
            "displayName": "Feedback",
            "propTypes": Object {
              "as": [Function],
              "type": [Function],
            },
            "render": [Function],
          },
          "defaultProps": Object {
            "as": "input",
          },
          "displayName": "FormControl",
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <SpeechRecognition
              setText={[Function]}
            />,
            <ForwardRef
              active={false}
              disabled={false}
              onClick={[Function]}
              type="button"
              variant="outline-primary"
            >
              <img
                alt="add.svg"
                className="inputicon"
                src="add.svg"
              />
            </ForwardRef>,
            <ForwardRef
              active={false}
              disabled={false}
              onClick={[Function]}
              type="button"
              variant="outline-danger"
            >
              <img
                alt="crossicon.svg"
                className="inputicon"
                src="crossicon.svg"
              />
            </ForwardRef>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "setText": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "active": false,
              "children": <img
                alt="add.svg"
                className="inputicon"
                src="add.svg"
              />,
              "disabled": false,
              "onClick": [Function],
              "type": "button",
              "variant": "outline-primary",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "add.svg",
                "className": "inputicon",
                "src": "add.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "active": false,
                "disabled": false,
                "type": "button",
                "variant": "primary",
              },
              "displayName": "Button",
              "render": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "active": false,
              "children": <img
                alt="crossicon.svg"
                className="inputicon"
                src="crossicon.svg"
              />,
              "disabled": false,
              "onClick": [Function],
              "type": "button",
              "variant": "outline-danger",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "crossicon.svg",
                "className": "inputicon",
                "src": "crossicon.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "active": false,
                "disabled": false,
                "type": "button",
                "variant": "primary",
              },
              "displayName": "Button",
              "render": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "displayName": "InputGroupAppend",
          "render": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "Append": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "displayName": "InputGroupAppend",
        "render": [Function],
      },
      "Checkbox": [Function],
      "Prepend": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "displayName": "InputGroupPrepend",
        "render": [Function],
      },
      "Radio": [Function],
      "Text": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "displayName": "InputGroupText",
        "render": [Function],
      },
      "defaultProps": undefined,
      "displayName": "Bootstrap(InputGroup)",
      "propTypes": undefined,
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef
            aria-describedby="basic-addon2"
            aria-label="Type task name"
            as="input"
            onChange={[Function]}
            onKeyPress={[Function]}
            placeholder="Type task name"
            size="lg"
            value=""
          />,
          <ForwardRef>
            <SpeechRecognition
              setText={[Function]}
            />
            <ForwardRef
              active={false}
              disabled={false}
              onClick={[Function]}
              type="button"
              variant="outline-primary"
            >
              <img
                alt="add.svg"
                className="inputicon"
                src="add.svg"
              />
            </ForwardRef>
            <ForwardRef
              active={false}
              disabled={false}
              onClick={[Function]}
              type="button"
              variant="outline-danger"
            >
              <img
                alt="crossicon.svg"
                className="inputicon"
                src="crossicon.svg"
              />
            </ForwardRef>
          </ForwardRef>,
        ],
        "className": "mb-3 mt-3",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "aria-describedby": "basic-addon2",
            "aria-label": "Type task name",
            "as": "input",
            "onChange": [Function],
            "onKeyPress": [Function],
            "placeholder": "Type task name",
            "size": "lg",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Feedback": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "as": "div",
                "type": "valid",
              },
              "displayName": "Feedback",
              "propTypes": Object {
                "as": [Function],
                "type": [Function],
              },
              "render": [Function],
            },
            "defaultProps": Object {
              "as": "input",
            },
            "displayName": "FormControl",
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <SpeechRecognition
                setText={[Function]}
              />,
              <ForwardRef
                active={false}
                disabled={false}
                onClick={[Function]}
                type="button"
                variant="outline-primary"
              >
                <img
                  alt="add.svg"
                  className="inputicon"
                  src="add.svg"
                />
              </ForwardRef>,
              <ForwardRef
                active={false}
                disabled={false}
                onClick={[Function]}
                type="button"
                variant="outline-danger"
              >
                <img
                  alt="crossicon.svg"
                  className="inputicon"
                  src="crossicon.svg"
                />
              </ForwardRef>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "setText": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "active": false,
                "children": <img
                  alt="add.svg"
                  className="inputicon"
                  src="add.svg"
                />,
                "disabled": false,
                "onClick": [Function],
                "type": "button",
                "variant": "outline-primary",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "add.svg",
                  "className": "inputicon",
                  "src": "add.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "active": false,
                  "disabled": false,
                  "type": "button",
                  "variant": "primary",
                },
                "displayName": "Button",
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "active": false,
                "children": <img
                  alt="crossicon.svg"
                  className="inputicon"
                  src="crossicon.svg"
                />,
                "disabled": false,
                "onClick": [Function],
                "type": "button",
                "variant": "outline-danger",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "crossicon.svg",
                  "className": "inputicon",
                  "src": "crossicon.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "active": false,
                  "disabled": false,
                  "type": "button",
                  "variant": "primary",
                },
                "displayName": "Button",
                "render": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "InputGroupAppend",
            "render": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "Append": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "displayName": "InputGroupAppend",
          "render": [Function],
        },
        "Checkbox": [Function],
        "Prepend": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "displayName": "InputGroupPrepend",
          "render": [Function],
        },
        "Radio": [Function],
        "Text": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "displayName": "InputGroupText",
          "render": [Function],
        },
        "defaultProps": undefined,
        "displayName": "Bootstrap(InputGroup)",
        "propTypes": undefined,
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
